core/src/main/java/org/commonjava/indy/core/inject/ExpiringMemoryNotFoundCache.java

```diff
@@ -52,12 +53,49 @@
     protected IndyConfiguration config;
 
     // TODO: Now using a simple hashmap, need to take attention here to see if need ISPN instead if it is a mem eater.
-    protected final Map<ConcreteResource, Long> missingWithTimeout = new HashMap<>();
+    protected final Map<ResourceSoftReference<ConcreteResource>, Long> missingWithTimeout = new HashMap<>();
```

Comment:
How is this functionally different from a WeakHashMap? Is the reference type difference important here?

IMO, implementing @pkocandr's idea of a parent-first check in the map (and avoiding adding child paths when the parent is marked as not found) might help control size here a bit more.

Reply:
@ruhan1 If we allow GC to expire these entries, that will short-circuit the NFC behavior, leading to Indy retrying the content retrieval earlier than it should. IMO this isn't what we want, since we could end up losing most or all of our NFC when memory pressure builds a little bit, and putting more pressure on the network and Indy itself by retrying all these downloads. This would make Indy less efficient as load increases.

We need the NFC to behave in a predictable way to preserve Indy performance. But we need it to expire things actively, not passively (see the `isMissing()` method).

Reply:
WeakReference (which is backing WeakHashmap) is different from SoftReference. AIUI, WeakReferences are cleaned when GC runs. GC runs quite often so the WeakHasmap often times retains very few entries or nothing at all. SoftReferences are cleaned only when the memory press is huge and an OOM is on the way. JVM makes sure it reclaims all soft-referenced objects before throws an OOM.